src/interpolation/details/ArborX_InterpDetailsPolynomialBasis.hpp

```diff
@@ -0,0 +1,173 @@
+//    |   | zz |      ||
+
+// This function returns the lengths of the slices for a given maximum degree
+// and dimension.
+// Given a slice at degree n and dimension u, its length would be the sum of all
+// the slices' lengths of degree n-1 and of same or smaller dimension.
+//
+// Given the previous two examples, the two arrays would be:
+// Deg \ Dim | x | y | z  || Deg \ Dim | x | y
+// ----------+---+---+--- || ----------+---+---
+//      1    | 1 | 1 | 1  ||      1    | 1 | 1
+//      2    | 1 | 2 | 3  ||      2    | 1 | 2
+//                        ||      3    | 1 | 3
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr auto polynomialBasisSliceLengths()
+{
+  static_assert(
+      Deg != 0 && Dim != 0,
+      "Unable to compute slice lengths for a constant polynomial basis");
+
+  struct
+  {
+    std::size_t arr[Deg][Dim]{};
+  } result;
+  auto &arr = result.arr;
+
+  for (std::size_t dim = 0; dim < Dim; dim++)
+    arr[0][dim] = 1;
+
+  for (std::size_t deg = 0; deg < Deg; deg++)
+    arr[deg][0] = 1;
+
+  for (std::size_t deg = 1; deg < Deg; deg++)
+    for (std::size_t dim = 1; dim < Dim; dim++)
+      arr[deg][dim] = arr[deg - 1][dim] + arr[deg][dim - 1];
+
+  return result;
+}
+
+// This returns the size of the polynomial basis, which is the sum of all the
+// slices lengths and 1.
+// Both of the previous examples will result in a polynomial basis of size 10
+template <std::size_t Dim, std::size_t Deg>
+KOKKOS_FUNCTION constexpr std::size_t polynomialBasisSize()
+{
+  if constexpr (Deg != 0 && Dim != 0)
+  {
+    auto [arr] = polynomialBasisSliceLengths<Dim, Deg>();
+    std::size_t size = 1;
+
+    for (std::size_t deg = 0; deg < Deg; deg++)
+      for (std::size_t dim = 0; dim < Dim; dim++)
+        size += arr[deg][dim];
+
+    return size;
+  }
+  else
+  {
+    return 1;
+  }
+}
+
+// This creates the list by building each slices in-place
+template <std::size_t Dim, std::size_t Deg, typename Point>
+KOKKOS_FUNCTION auto evaluatePolynomialBasis(Point const &p)
+{
+  using value_t = std::decay_t<decltype(p[0])>;
+  Kokkos::Array<value_t, polynomialBasisSize<Dim, Deg>()> arr{};
+  arr[0] = value_t(1);
+
+  if constexpr (Deg != 0 && Dim != 0)
+  {
+    // Cannot use structured binding with constexpr
+    static constexpr auto slice_lengths_struct =
+        polynomialBasisSliceLengths<Dim, Deg>();
+    static constexpr auto &slice_lengths = slice_lengths_struct.arr;
+
+    std::size_t prev_col = 0;
+    std::size_t curr_col = 1;
+
+    for (std::size_t deg = 0; deg < Deg; deg++)
+    {
+      std::size_t loc_offset = curr_col;
+      for (std::size_t dim = 0; dim < Dim; dim++)
+      {
+        // copy the previous column and multply by p[dim]
+        for (std::size_t i = 0; i < slice_lengths[deg][dim]; i++)
+          arr[loc_offset + i] = arr[prev_col + i] * p[dim];
+
+        loc_offset += slice_lengths[deg][dim];
+      }
+
+      prev_col = curr_col;
+      curr_col = loc_offset;
+    }
```

Comment:
I wonder if a compiler is smart enough to optimize this code. That is a lot of for loops and calls if we want to simply evaluate a quadratic polynomial in 2D or 3D. Wonder if we should specialize. Not blocking, though.

Reply:
If this part is slow, unrolling might help as the loop sizes are known at compile time and would be somewhat small.